### PR TITLE
Resolve scalar overloads before AOT compilation

### LINKED ISSUE #90923: dynamo.optimizations.training.aot_autograd does not trace correct overload
### 🐛 Describe the bug

In the following Python program, `x + 2` gets traced as `torch.ops.aten.add.Tensor` instead of `torch.ops.aten.add.Scalar` which would be more technically correct (and is the op that torch.jit.script gives). I noticed this because in Torch-MLIR we verify that the operands match the schema exactly. Is it possible to have aot_autograd trace the more technically correct overload? Or can you provide guidance about how to correctly emulate the promotion semantics in appropriate generality?

Also, to add onto that, `torch.ops.aten.add.Tensor` is missing an argument. I need special-case code in my importer to handle this (by querying the schema for the default value). One possibility could be that PyTorch provides a utility that normalizes the graph of aten ops to "typecheck" correctly against the schemas of the constituent ops including reifying all default values.

```python
from typing import List

import torch
import torch._dynamo as dynamo
from torch._dynamo.optimizations.training import aot_autograd


def my_backend(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    gm.print_readable()
    return gm


@dynamo.optimize(aot_autograd(fw_compiler=my_backend))
def f(x):
    return x + 2


example_inputs = (torch.randn(3, 4),)
f(*example_inputs)
```

Output:
```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3, 4]):
        # File: /tmp/repro2.py:15, code: return x + 2
        add: f32[3, 4] = torch.ops.aten.add.Tensor(arg0_1, 2);  arg0_1 = None
        return (add,)     
```

### Versions

PyTorch version: 2.0.0.dev20221213+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux rodete (x86_64)
GCC version: (Debian 12.2.0-3) 12.2.0
Clang version: 14.0.6-2
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.8 (main, Nov  3 2022, 15:17:13) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.19.11-1rodete1-amd64-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.0rc2
[pip3] torch==2.0.0.dev20221213+cpu
[pip3] torchvision==0.15.0.dev20221213+cpu
[conda] Could not collect

cc @chauhang @penguinwu @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4 @bdhirsh @bobrenjc93 @aorenste @ezyang @anijain2305 @gmagogsfm @zou3519 @msaroufim @wconstab @soumith @ngimel

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186557

AOTAutograd can hand backend compilers FX graphs whose node target is a Tensor overload even though the graph still contains a Python scalar literal, for example aten.add.Tensor(x, 2). That mirrors what the Python arg parser sent to the dispatcher for operators that allow numbers as tensors, but it is schema-inexact for backends that typecheck ATen operators directly.

Fix this at the AOT boundary by walking the produced forward/backward GraphModules, including nested HOP subgraphs, and resolving ATen packet overloads against the actual FX argument values. The pass only rewrites when the packet has a unique matching overload, so tensor/tensor cases such as torch.add(x, y, alpha=2) stay on the Tensor overload. It also handles symbolic Python scalars and tensorifies scalar-left floor_divide where no scalar-left overload exists.

Because Inductor sees these corrected scalar overloads, update the affected binary folding, DDP fusion, and SDPA pattern matching consumers to accept Tensor and Scalar scale variants. Regenerate the serialized SDPA patterns and teach pattern serialization how to represent multi-overload targets for these scalar workaround inputs.

I chose overload re-resolution at the AOT compiler boundary instead of backend-specific importer rewrites because it keeps all downstream backend graphs schema-accurate and delegates overload choice to existing operator schema information.

Benchmark Results

A direct before/after compile benchmark against a clean main worktree could not run because that auxiliary worktree was not built and could not import torch. On the built patched checkout, compiling a 20-iteration scalar arithmetic AOT function took median 0.103263s over 8 runs. A microbenchmark of the new overload pass on a 60-op FX graph measured baseline deepcopy median 0.001667s and pass-only median 0.003516s, mean 0.003548s.

Test Plan

python test/dynamo/test_backends.py TestCustomBackendAPI

python test/inductor/test_pattern_matcher.py TestPatternMatcher.test_serialized_patterns_up_to_date

python test/inductor/test_fused_attention.py SDPAPatternRewriterCpuTests.test_sdpa_rewriter_1_cpu SDPAPatternRewriterCpuDynamicTests.test_sdpa_rewriter_1_cpu

python test/inductor/test_fused_attention.py SDPAPatternRewriterCpuTests.test_sdpa_rewriter_2_cpu SDPAPatternRewriterCpuTests.test_sdpa_rewriter_12_cpu

python test/dynamo/test_base_hop.py BaseHOPTest.test_auto_functionalize && python test/dynamo/test_graph_deduplication.py GraphDededuplicationTests.test_input_aliasing GraphDededuplicationTests.test_autocast_ordering && python test/dynamo/test_regional_inductor.py -k 'repeated_blocks' && python test/dynamo/test_regional_inductor.py -k 'invoke_subgraph_inner' && python test/higher_order_ops/test_invoke_subgraph.py TestInvokeSubgraphCompile.test_auto_functionalize

python test/inductor/test_binary_folding.py -k 'binary_folding' with a local torchvision stub due to the environment's torchvision::nms collection failure

python test/functorch/test_aotdispatch.py TestAOTAutograd with a local torchvision stub due to the environment's torchvision::nms collection failure

lintrunner -a $(git diff --name-only)

git diff --check

Fixes #90923

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR, fixed the scalar-left symbolic floor_divide bug caught during review, and updated stale scalar-overload expectations behind the CI failures.

**Comment by jansel** [comment]
My agent says I addressed the Dr. CI scalar-overload fallout by updating the remaining stale add/mul Scalar graph expectations and the MKLDNN SiLU matcher node-count expectations.

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the Dr. CI failures by updating stale Scalar overload graph expectations and restoring the MKLDNN SiLU matcher node-count expectations. The named failing tests and lintrunner pass locally; CI is now rerunning.

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the Dr. CI scalar-overload fallout: remove_no_ops now handles add/sub/mul/div Scalar, binary folding preserves MKLDNN hardsigmoid fusion, and stale nested graph expectations are updated. The targeted failing tests and lintrunner pass locally; CI is rerunning.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the current Dr. CI fallout: stale Scalar overload expectations in the nested graph and opaque-object tests now match the schema-accurate AOTAutograd graphs, and the scalar_tensor metadata / SDPA pattern checks have small robustness cleanups. The targeted failing tests, related AOT/Inductor checks, and lintrunner pass locally; CI is rerunning.

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the current Dr. CI failures by refreshing the stale Scalar overload inline expectations in graph deduplication, invoke-subgraph, and opaque-object tests. The named failing tests, the AOTInductor logging check with a local torchvision stub, and lintrunner pass locally; CI is rerunning.

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the current Dr. CI failure by updating the stale opaque-object expected graph from aten.add.Tensor to aten.add.Scalar for the scalar literal case. The failing test plus the related graph-dedup/subclass checks and lintrunner pass locally; CI is rerunning.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
